runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java

```diff
@@ -111,6 +112,25 @@ public void translate(final RunnerApi.Pipeline pipeline, SparkTranslationContext
     QueryablePipeline p =
         QueryablePipeline.forTransforms(
             pipeline.getRootTransformIdsList(), pipeline.getComponents());
+    for (PipelineNode.PTransformNode transformNode : p.getTopologicallyOrderedTransforms()) {
+      // Pre-scan pipeline to count which pCollections are consumed as inputs more than once so
+      // their corresponding RDDs can later be cached.
+      for (String inputId : transformNode.getTransform().getInputsMap().values()) {
+        context.incrementConsumptionCountBy(inputId, 1);
```

Comment:
I was thinking that rather than putting consumption counts on the context, we would just compute them locally and then inform the context of the full set of collections that need caching (together with their coders). But we can keep it this way if you'd prefer.

Reply:
Ack, I think I'm just going to leave it the way it is -- it somehow feels better to pass integers than maps. :man_shrugging: